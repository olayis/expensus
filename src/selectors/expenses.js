// Get Visible expenses
import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, 'day')
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, 'day')
        : true;
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return textMatch && startDateMatch && endDateMatch;
    })
    .sort((a, b) => {
      switch (sortBy) {
        case 'date(recent)':
          return a.createdAt < b.createdAt ? 1 : -1;

        case 'date(oldest)':
          return a.createdAt > b.createdAt ? 1 : -1;

        case 'amount(highest)':
          return a.amount < b.amount ? 1 : -1;

        case 'amount(lowest)':
          return a.amount > b.amount ? 1 : -1;

        case 'description(a_to_z)':
          return a.description.localeCompare(b.description);

        case 'description(z_to_a)':
          return b.description.localeCompare(a.description);

        case 'category':
          return a.category.localeCompare(b.category);
      }
    });
};
