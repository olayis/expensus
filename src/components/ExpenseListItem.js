import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import moment from 'moment';
import numeral from 'numeral';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faMoneyBillAlt,
  faCalendarAlt,
} from '@fortawesome/free-regular-svg-icons';
import { faTags, faTrash } from '@fortawesome/free-solid-svg-icons';
import { startRemoveExpense } from '../actions/expenses';
import ConfirmationModal from './ConfirmationModal';

export class ExpenseListItem extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      modalSelected: false,
    };
  }

  onRemove = () => {
    this.props.startRemoveExpense({ id: this.props.id });
  };

  toggleModalSelection = () => {
    this.setState((prevState) => ({
      modalSelected: !prevState.modalSelected,
    }));
  };

  render() {
    return (
      <>
        <div className='card-box'>
          <button
            className='card-box__delete'
            onClick={this.toggleModalSelection}
            aria-label='Remove expense'
          >
            <FontAwesomeIcon className='fa-icon' icon={faTrash} />
          </button>
          <Link className='router-link' to={`edit/${this.props.id}`}>
            <h3 className='card-box__description'>{this.props.description}</h3>
            <p className='card-box__amount'>
              <FontAwesomeIcon className='fa-icon' icon={faMoneyBillAlt} />
              {numeral(this.props.amount / 100).format('$0,0.00')}
            </p>
            <p className='card-box__category'>
              <FontAwesomeIcon className='fa-icon' icon={faTags} />
              {this.props.category}
            </p>
            <p className='card-box__date'>
              <FontAwesomeIcon className='fa-icon' icon={faCalendarAlt} />
              {moment(this.props.createdAt).format('MMM Do, YYYY')}
            </p>
          </Link>
        </div>
        <ConfirmationModal
          handleModalAction={this.onRemove}
          modalSelected={this.state.modalSelected}
          cancelModalSelection={this.toggleModalSelection}
          modalTitle='Confirm Delete'
          modalDescription={`Delete ${this.props.description}`}
          actionButtonLabel='Delete'
        />
      </>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data)),
});

export default connect(undefined, mapDispatchToProps)(ExpenseListItem);
