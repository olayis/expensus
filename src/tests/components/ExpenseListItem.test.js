import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import ExpenseListItem from '../../components/ExpenseListItem';

test('should render ExpenseListItem with expense', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[2]} />);
  expect(wrapper).toMatchSnapshot();
});
