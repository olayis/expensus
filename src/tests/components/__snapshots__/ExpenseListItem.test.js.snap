// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with expense 1`] = `
<div>
  <Link
    to="edit/3"
  >
    <h3>
      Electricity
    </h3>
  </Link>
  <p>
    8000
  </p>
  <p>
    <Component />
  </p>
</div>
`;
